server/src/main/java/org/elasticsearch/action/admin/indices/mapping/get/GetFieldMappingsResponse.java

```diff
@@ -20,27 +20,58 @@
 package org.elasticsearch.action.admin.indices.mapping.get;
 
 import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.common.ParseField;
 import org.elasticsearch.common.bytes.BytesArray;
 import org.elasticsearch.common.bytes.BytesReference;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ObjectParser;
 import org.elasticsearch.common.xcontent.ToXContentFragment;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.common.xcontent.XContentParser;
 import org.elasticsearch.common.xcontent.XContentType;
 import org.elasticsearch.index.mapper.Mapper;
 
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Objects;
 
 import static java.util.Collections.emptyMap;
 import static java.util.Collections.unmodifiableMap;
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.optionalConstructorArg;
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
 
 /** Response object for {@link GetFieldMappingsRequest} API */
 public class GetFieldMappingsResponse extends ActionResponse implements ToXContentFragment {
 
+    private static final ParseField MAPPINGS = new ParseField("mappings");
+
+    private static final ObjectParser<Map<String, Map<String, FieldMappingMetaData>>, String> PARSER =
+        new ObjectParser<>(MAPPINGS.getPreferredName(), true, HashMap::new);
+
+    static {
+        PARSER.declareField((p, typeMappings, index) -> {
+            while (p.nextToken() == XContentParser.Token.FIELD_NAME) {
+                final String typeName = p.currentName();
+                final Map<String, FieldMappingMetaData> typeMapping = new HashMap<>();
+                typeMappings.put(typeName, typeMapping);
+
+                if (p.nextToken() == XContentParser.Token.START_OBJECT) {
```

Comment:
So, if `nextToken` is `START_ARRAY` I think this falls over. I think the outer while loop terminates in a weird spot. I think we should skip children in that case.